builtin/logical/pki/cert_util.go

```diff
@@ -1513,3 +1514,371 @@ func stringToOid(in string) (asn1.ObjectIdentifier, error) {
+			// as a heuristic for whether or not this relationship is valid,
+			// this is insufficient as otherwise valid CA certificates could
+			// elide this information. That means its best to actually validate
+			// the signature (e.g., call child.CheckSignatureFrom(parent))
+			// instead.
+			parentCert := issuerIdCertMap[parent]
+			if err := childCert.CheckSignatureFrom(parentCert); err != nil {
+				return nil
+			}
+
+			// Otherwise, we can append it to the map, allowing us to walk the
+			// issuer->parent mapping.
+			issuerIdParentsMap[child] = append(issuerIdParentsMap[child], parent)
+
+			// Also cross-add the child relationship step at the same time.
+			issuerIdChildrenMap[parent] = append(issuerIdChildrenMap[parent], child)
+		}
+	}
+
+	// Finally, we consult RFC 8446 Section 4.4.2 for creating an algorithm for
+	// building the chain:
+	//
+	// > ... The sender's certificate MUST come in the first
+	// > CertificateEntry in the list.  Each following certificate SHOULD
+	// > directly certify the one immediately preceding it.  Because
+	// > certificate validation requires that trust anchors be distributed
+	// > independently, a certificate that specifies a trust anchor MAY be
+	// > omitted from the chain, provided that supported peers are known to
+	// > possess any omitted certificates.
+	// >
+	// > Note: Prior to TLS 1.3, "certificate_list" ordering required each
+	// > certificate to certify the one immediately preceding it; however,
+	// > some implementations allowed some flexibility.  Servers sometimes
+	// > send both a current and deprecated intermediate for transitional
+	// > purposes, and others are simply configured incorrectly, but these
+	// > cases can nonetheless be validated properly.  For maximum
+	// > compatibility, all implementations SHOULD be prepared to handle
+	// > potentially extraneous certificates and arbitrary orderings from any
+	// > TLS version, with the exception of the end-entity certificate which
+	// > MUST be first.
+	//
+	// So, we take this to mean we should build chains via DFS: each issuer is
+	// explored until an empty parent pointer (i.e., self-loop) is reached and
+	// then latest-most duplicate parent is then explored.
+	//
+	// However, we don't actually need to do a DFS (per issuer) here. We can
+	// simply invert the (pseudo-)directed graph, i.e., topologically sort it.
+	// Some number of certs (roots without cross-signing) lack parent issuers.
+	// These are already "done" from the PoV of chain building. We can thus
+	// iterating through the parent mapping to find entries without parents to
+	// start the sort. After processing, we can add all children and visit them
+	// if all parents have been processed.
+	//
+	// Note though, that while topographical sorting is equivalent to the DFS,
+	// we have to take care to make it a pseudo-DAG. This means handling the
+	// most common 2-star (2-clique) sub-graphs of reissued certificates,
```

Comment:
Presently we only handle cliques as the roots of trust. I think this is incorrect. In particular, consider the restrictions under which cliques form:

 - The subject and issuer of all certs in the clique needs to be the same, in order to generate the clique.
 - The signature needs to be valid for the parent relationship to exist. This means that all members of the clique need to use the same key material.

In particular, if there were another key using the same key material with a different Subject or Issuer, it would not be the part of the clique. If there were another key using the different key material with the same subject, it could not be in the clique as the signature (onto members of the clique) would be invalid. 

However, we do have to consider cross-signed certs (same subject as the clique but not itself part of the clique -- due to different issuer). Because of the different issuer they'd allow validation of the clique (from an non-clique root of trust), effectively making the clique an internal/terminal clique (from the purposes of validation).  

Since cliques are formed by reissuance of certs, this would concretely look like an old root being reissued (at least once, if not more). During rotation to a new key material for a new root, the old root could be cross-signed by the new root. This forms a clique (by reissuance of the old) that's not the root of trust (if the new root is imported into the mount). That suggests we probably need to handle the scenario where cliques aren't roots of trust, and thus refactor it into a method, to be called from within the `toVisit` queue. In particular, we wish to visit the clique after all external potential references into it have been processed, to ensure the chain is constructed correctly. Bounding clique size probably helps. 

This clique detection also detects two-cycles (as referenced by the circle cert example). We don't detect any larger cycles (as cycles are truly pathological and do not provide much value, if any unlike cliques which form from reissuance). Maybe the scenario where the old root cross-signs the new root, in which case we'd get a non-clique cycle of (old root (self-sig) -> new root crossed by old root -> new root (self sig) -> new root cross old root -> [repeat])

So, I think the net is that we need to go back and revise this clique detection logic and separately also add cycle detection with cross-signed roots. In handling the cycle detection, we'll still want to include all parts of the cycle (including the loop of new root back to the present value), as that'll allow devices using the new cert to validate the old root without needing to use a different issuer. Cycles need not be at the root of trust either.

Yikes!